modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1837,18 +1821,20 @@ public CompletableFuture<List<Table>> tablesAsync() {
      * @return Future representing pending completion of the operation.
      */
     private CompletableFuture<List<Table>> tablesAsyncInternal() {
-        return supplyAsync(() -> inBusyLock(busyLock, this::directTableIds), ioExecutor)
-                .thenCompose(tableIds -> inBusyLock(busyLock, () -> {
-                    var tableFuts = new CompletableFuture[tableIds.size()];
+        return schemaSyncService.waitForMetadataCompleteness(clock.now())
```

Comment:
`TableManager#tables` may hang, and if you try to stop the `TableManager`, you will permanently hang the stop of the node.

See how I fixed this problem.
https://github.com/apache/ignite-3/blob/c63c4530f00ff67895c734e9221eac1189849d78/modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java#L1444C26-L1444C26

Reply:
However, I believe, we should never do `future.join()` inside the busylock.

```
 private <T> T join(CompletableFuture<T> future) {
        if (!busyLock.enterBusy()) {
            throw new IgniteException(new NodeStoppingException());
        }

        try {
            return future.join();
        } catch (CompletionException ex) {
            throw convertThrowable(ex.getCause());
        } finally {
            busyLock.leaveBusy();
        }
    }
```

Reply:
On the one hand, you are right, but on the other hand, they won’t do it in another place. 
In theory, if a component is stopped, then it is responsible to complete the future with an error, and not rely on another component.